Add build trigger options, pull templates from installed pkg

k-doering-NOAA opened the pull request:
Closes #33 . 

- Adds more build trigger options to the R functions
- Includes tests for the new build trigger options
- Pulls template caller workflow files from the installed ghactions4r package instead of from main. This will ensure that the templates are designed to work with the R function

This is based on https://github.com/nmfs-ost/ghactions4r/pull/220, but applied to all relevant workflows ([details](https://github.com/nmfs-ost/ghactions4r/issues/33#issuecomment-3156807507)).

Two open questions:
- There is one additional open issue to allow for multiple build triggers (#135). This PR only allows users to select a SINGLE build trigger, but if desired, I can also address the multiple build triggers work in this PR.
- I used `@template` in the roxygen documentation, though it is an out-of-date method, as I could not quickly comprehend what to switch to. However, I plan to address this in the future using #224 - if necessary, I can switch before this PR gets merged in, though.

k-doering-NOAA committed:
Refactor: Change how to get caller workflow
Get the template caller workflow from the
downloaded pkg itself rather than the current
main branch of ghactions4r.

k-doering-NOAA committed:
refactor: build triggers
separate out build trigger function,
use in doc and style r, update_pkgdown

k-doering-NOAA committed:
Download template from pkg instead of main

k-doering-NOAA committed:
fix test and update snaps

k-doering-NOAA committed:
fix: accidentally removed code
during rebase merge conflicts resolution

k-doering-NOAA committed:
Add build triggers, change to using copy_caller_template

instead of downloading template files from the main branch of GitHub.

k-doering-NOAA committed:
Add missing tests

Other build triggers covered well by doc and style, but was missing the push_to_all_branches option.

k-doering-NOAA committed:
fix devtools::check and update documentation
rm octocov template from Rbuildignore

k-doering-NOAA commented:
@Bai-Li-NOAA , would you be willing to review by 9/15?

k-doering-NOAA commented:
Hi @Bai-Li-NOAA, I think your week has been unexpectedly busy! A number of the NSAP team members have volunteered to help out with ghactions4r pull request reviews after I asked for help with this at our Monday standup. Would you like me to ask someone else to review this PR?

Bai-Li-NOAA commented on tests/testthat/_snaps/use_r_workflows.md:
```diff
@@ -7,25 +7,20 @@
-      [23] "      depends_on_quarto: true"                                                                                                            
+      [12] "jobs:"                                                                                                                                    
+      [13] "  call-workflow:"                                                                                                                         
+      [14] "    uses: nmfs-ost/ghactions4r/.github/workflows/r-cmd-check.yml@main"                                                                    
```
One thing I don’t fully understand is that the R function pulls template caller workflow files from the installed ghactions4r package instead of from main, but the caller workflow YAML file still uses the workflow from main. Could this cause compatibility issues? For example, some parameters under the with section don’t work with the workflow YAML file from main.

If a user is working with version v0.9 of the R functions, do they need to use: `uses: nmfs-ost/ghactions4r/.github/workflows/r-cmd-check.yml@v0.9`? Or am I overthinking this?

k-doering-NOAA replied:
@Bai-Li-NOAA, this is a really good point! This is a real concern.

I think if a user *absolutely* wanted to ensure compatibility, it is correct that they would need to pin to the version of the package they are using locally.  However, we typically add variables and not take them away so far, so I think the chances of this compatibility issue are low.

However, pinning to versions would mean users would need to update their YAML caller workflows periodically, which to me doesn't seem worthwhile the tradeoff for a user who isn't comfortable in GitHub Actions. 

What are your thoughts on the path we should take forward?

Bai-Li-NOAA replied:
@k-doering-NOAA, that makes sense. I agree the risk is likely low since we’ve generally added parameters rather than removed them. One option could be to keep `@main` as the default for ease of use, and add a note in the documentation that pinning to a specific version is mainly for users who need strict reproducibility (with the tradeoff that they’ll need to update their YAML caller workflows periodically).

Another option could be to add compatibility tests that check the installed `{ghactions4r}` package (e.g., the last X versions) against the `@main` workflows. Alternatively, we could simply rely on debugging mismatches as they come up, since the GHA logs include the commit number associated with each run.

k-doering-NOAA replied:
I kept main as the default, but added an option to add tags from ghactions4r to the r functions. The documentation also includes how to pin to specific commits, but notes that users will have to do that manually after generating the file.

I think for now, debugging mismatches as they come up seems fine. If it begins to be a big problem, we can add compatibility tests, I think.

Bai-Li-NOAA commented on R/use_r_workflows.R:
```diff
@@ -28,26 +29,26 @@
 #' }
 #' @export
 use_r_cmd_check <- function(workflow_name = "call-r-cmd-check.yml",
```
The function returns the following error when a `call-r-cmd-check.yml` file already exists in the repo: `Error in if (!after) c(values, x) else if (after >= lengx) c(x, values) else c(x[1L:after],  : 
  argument is of length zero`. Maybe the function need to overwrite the existing workflow and notify users in the console? This may also apply to the other R functions.

k-doering-NOAA replied:
AH, thank you! This is a case I did not check, apparently!!!

I will add a tests and make a fix next week.

Bai-Li-NOAA commented on R/use_r_workflows.R:
```diff
@@ -28,26 +29,26 @@
 #' }
 #' @export
 use_r_cmd_check <- function(workflow_name = "call-r-cmd-check.yml",
+                            build_trigger = c("manually", "weekly", "pull_request", "push_to_main", "push_to_all_branches"),
```
Do you know which option is most commonly used by users? Would `push_to_main` be an easier default? Is it used more often? I have the same question regarding the order of build triggers in the other functions.

k-doering-NOAA replied:
Oh, this is an excellent point! I think it might differ by workflow, but manually is not likely a good default in all cases!

I will work on this next week.

Bai-Li-NOAA commented on R/utils.R:
```diff
@@ -70,3 +70,55 @@ add_quarto_true <- function(uses_line, gha) {
+#' @param gha The workflow file that has been read in using readLines
+#' @return The modified workflow file
+#' @noRd
+add_build_trigger <- function(build_trigger, gha) {
```
When testing locally, it seems that a user can only select one `build_trigger` at a time. For example, running `use_r_cmd_check(build_trigger = c("weekly", "push_to_main"))` returns the error: "in match.arg(arg = build_trigger) : 'arg' must be of length 1". Would you like the function to support multiple build triggers as a vector, or could the description for `build_trigger` be updated to clarify that only one trigger can be provided?

k-doering-NOAA replied:
I think for this PR, I will add the error message about one build trigger only, and then will edit it when working on https://github.com/nmfs-ost/ghactions4r/issues/135

Bai-Li-NOAA reviewed (commented):
Thanks @k-doering-NOAA for making the R functions more flexible. They look good. I’ve done a few edge case tests locally and added some questions for consideration.

Bai-Li-NOAA commented:
@k-doering-NOAA, it’s great to hear that more people are available to help with PR reviews! 🥳  I’ve finished reviewing this PR. Sorry for the delay, my little one has been sick at home, and I just got a chance to look at the PR today.

Regarding the two open questions, I like keeping each PR focused on a single task, as it makes reviewing and testing easier. Addressing the other issues in separate PRs works for me.

k-doering-NOAA commented:
Thanks @Bai-Li-NOAA for taking a look at this, I really appreciate your time! I will look at your comments next week.

k-doering-NOAA committed:
add to WORDLIST

k-doering-NOAA committed:
add a human readable error for multiple triggers
Multiple build triggers are not yet possible, so add a error
message using cli for now.

k-doering-NOAA committed:
Reorder build triggers
put most likely used options first
suggested by @Bai-Li-NOAA

k-doering-NOAA committed:
fix: old workflows can be overwritten

k-doering-NOAA committed:
Add option for adding tag instead of main
In caller workflows. Suggested by @Bai-Li-NOAA

k-doering-NOAA committed:
spelling fixes
Add to wordlist, fix some errors.

k-doering-NOAA committed:
Add a pull request template
To document who is available to review ghactions4r.

k-doering-NOAA commented:
@Bai-Li-NOAA , I have finished resolving your suggestions - would you be able to take a look at the new code and see if it meets your suggestions?

I also added a pull request template that just documents who can review ghactions4r pull requests in the future - I can add to a separate branch instead if it is a nuisance! 

Let me know if you think this is ready to merge in or not.

Bai-Li-NOAA commented on R/utils.R:
```diff
@@ -70,3 +70,108 @@ add_quarto_true <- function(uses_line, gha) {
+  available_tags <- get_tags(repo = repo, headers = NULL)
+  if(tag %in% available_tags) {
+    # sub
+      gha <- sub("@main", paste0("@", tag), gha, fixed = TRUE)
```
I think this should work for all existing workflows. Just a note: if a workflow contains multiple `@main` references, only the first one will be replaced.

k-doering-NOAA replied:
good point, I saw that, but I think there should be just a single reference to main per workflow!

Bai-Li-NOAA commented:
> @Bai-Li-NOAA , I have finished resolving your suggestions - would you be able to take a look at the new code and see if it meets your suggestions?
> 
> I also added a pull request template that just documents who can review ghactions4r pull requests in the future - I can add to a separate branch instead if it is a nuisance!
> 
> Let me know if you think this is ready to merge in or not.

@k-doering-NOAA, thanks for addressing all the suggestions! I just have a minor note and the PR is ready to merge. I also reviewed the PR template, and it looks great.

k-doering-NOAA commented:
Thanks @Bai-Li-NOAA , I'm going to merge this in 🎉! Thank you for all your time reviewing changes, I really appreciate it.

Note: I'm excited that there is now only 1 remaining issue left on the v1.0.0 milestone!
